src/main/drivers/camera_control.c

```diff
@@ -0,0 +1,217 @@
+        #else
+        IOConfigGPIO(cameraControlPwm.io, IOCFG_AF_PP);
+        #endif
+
+        pwmOutConfig(&cameraControlPwm, timerHardware, CAMERA_CONTROL_TIMER_MHZ, CAMERA_CONTROL_PWM_RESOLUTION, 0, 0);
+
+        *cameraControlPwm.ccr = cameraControlPwm.period;
+        cameraControlPwm.enabled = true;
+    } else if (CAMERA_CONTROL_MODE_SOFTWARE_PWM == cameraControlConfig()->mode) {
+#ifdef CAMERA_CONTROL_SOFTWARE_PWM_AVAILABLE
+        IOConfigGPIO(cameraControlPwm.io, IOCFG_OUT_PP);
+        IOHi(cameraControlPwm.io);
+
+        cameraControlPwm.period = CAMERA_CONTROL_SOFT_PWM_RESOLUTION;
+        cameraControlPwm.enabled = true;
+
+        NVIC_InitTypeDef nvicTIM6 = {
+            TIM6_DAC_IRQn, NVIC_PRIORITY_BASE(NVIC_PRIO_TIMER), NVIC_PRIORITY_SUB(NVIC_PRIO_TIMER), ENABLE
+        };
+        NVIC_Init(&nvicTIM6);
+        NVIC_InitTypeDef nvicTIM7 = {
+            TIM7_IRQn, NVIC_PRIORITY_BASE(NVIC_PRIO_TIMER), NVIC_PRIORITY_SUB(NVIC_PRIO_TIMER), ENABLE
+        };
+        NVIC_Init(&nvicTIM7);
+
+        RCC->APB1ENR |= RCC_APB1Periph_TIM6 | RCC_APB1Periph_TIM7;
+        TIM6->PSC = 0;
+        TIM7->PSC = 0;
+#endif
+    } else if (CAMERA_CONTROL_MODE_DAC == cameraControlConfig()->mode) {
+        // @todo not yet implemented
+    }
+}
+
+void cameraControlProcess(uint32_t currentTimeUs)
+{
+    if (endTimeMillis && currentTimeUs >= 1000 * endTimeMillis) {
+        if (CAMERA_CONTROL_MODE_HARDWARE_PWM == cameraControlConfig()->mode) {
+            *cameraControlPwm.ccr = cameraControlPwm.period;
+        } else if (CAMERA_CONTROL_MODE_SOFTWARE_PWM == cameraControlConfig()->mode) {
+
+        }
+
+        endTimeMillis = 0;
+    }
+}
+
+static const int cameraPullUpResistance = 47000;
+static const int buttonResistanceValues[] = { 45000, 27000, 15000, 6810, 0 };
+
+static float calculateKeyPressVoltage(const cameraControlKey_e key)
+{
+    const int buttonResistance = buttonResistanceValues[key];
+    return 1.0e-2f * cameraControlConfig()->refVoltage * buttonResistance / (cameraPullUpResistance + buttonResistance);
+}
+
+static float calculatePWMDutyCycle(const cameraControlKey_e key)
+{
+    const float voltage = calculateKeyPressVoltage(key);
+
+    return voltage / ADC_VOLTAGE;
+}
+
+void cameraControlKeyPress(cameraControlKey_e key)
+{
+    if (key >= CAMERA_CONTROL_KEYS_COUNT)
+        return;
+
+    const float dutyCycle = calculatePWMDutyCycle(key);
+
+    if (CAMERA_CONTROL_MODE_HARDWARE_PWM == cameraControlConfig()->mode) {
+        *cameraControlPwm.ccr = lrintf(dutyCycle * cameraControlPwm.period);
+        endTimeMillis = millis() + cameraControlConfig()->keyDelayMs;
+    } else if (CAMERA_CONTROL_MODE_SOFTWARE_PWM == cameraControlConfig()->mode) {
+#ifdef CAMERA_CONTROL_SOFTWARE_PWM_AVAILABLE
+        const uint32_t hiTime = lrintf(dutyCycle * cameraControlPwm.period);
+
+        if (0 == hiTime) {
+            IOLo(cameraControlPwm.io);
```

Comment:
Most IO functions work correctly with io==NONE (doing nothing). Pin initialization may be an exception